django_lightweight_queue/app_settings.py

```diff
@@ -1,53 +1,90 @@
-from typing import Dict, Union, Mapping, TypeVar, Callable, Optional, Sequence
+from typing import Any, Dict, List, Union, Callable, Optional, Sequence
 
-from django.conf import settings
+from typing_extensions import Protocol
+
+from django.conf import settings as django_settings
 
 from . import constants
 from .types import Logger, QueueName
 
-T = TypeVar('T')
 
+class LongNameAdapter:
+    def __init__(self, target: Any) -> None:
+        self.target = target
+
+    def __getattr__(self, name: str) -> Any:
+        return getattr(self.target, f'{constants.SETTING_NAME_PREFIX}{name}')
+
+
+class Settings(Protocol):
+    WORKERS: Dict[QueueName, int]
+    BACKEND: str
+    LOGGER_FACTORY: Union[str, Callable[[str], Logger]]
+
+    # Allow per-queue overrides of the backend.
+    BACKEND_OVERRIDES: Dict[QueueName, str]
+    MIDDLEWARE: Sequence[str]
+
+    # Apps to ignore when looking for tasks. Apps must be specified as the dotted
+    # name used in `INSTALLED_APPS`. This is expected to be useful when you need to
+    # have a file called `tasks.py` within an app, but don't want
+    # django-lightweight-queue to import that file.
+    # Note: this _doesn't_ prevent tasks being registered from these apps.
+    IGNORE_APPS: Sequence[str]
+
+    REDIS_HOST: str
+    REDIS_PORT: int
+    REDIS_PASSWORD: Optional[str]
+    REDIS_DATABASE: int
+    REDIS_PREFIX: str
+
+    ENABLE_PROMETHEUS: bool
+    # Workers will export metrics on this port, and ports following it
+    PROMETHEUS_START_PORT: int
+
+    ATOMIC_JOBS: bool
+
+
+class Defaults(Settings):
+    WORKERS: Dict[QueueName, int] = {}
+    BACKEND = 'django_lightweight_queue.backends.synchronous.SynchronousBackend'
+    LOGGER_FACTORY = 'logging.getLogger'
+
+    BACKEND_OVERRIDES: Dict[QueueName, str] = {}
+    MIDDLEWARE = ('django_lightweight_queue.middleware.logging.LoggingMiddleware',)
+
+    IGNORE_APPS: Sequence[str] = ()
+
+    REDIS_HOST = '127.0.0.1'
+    REDIS_PORT = 6379
+    REDIS_PASSWORD = None
+    REDIS_DATABASE = 0
+    REDIS_PREFIX = ""
 
-def setting(suffix: str, default: T) -> T:
-    attr_name = '{}{}'.format(constants.SETTING_NAME_PREFIX, suffix)
-    return getattr(settings, attr_name, default)
+    ENABLE_PROMETHEUS = False
 
+    PROMETHEUS_START_PORT = 9300
 
-WORKERS = setting('WORKERS', {})  # type: Dict[QueueName, int]
-BACKEND = setting(
-    'BACKEND',
-    'django_lightweight_queue.backends.synchronous.SynchronousBackend',
-)  # type: str
+    ATOMIC_JOBS = True
 
-LOGGER_FACTORY = setting(
-    'LOGGER_FACTORY',
-    'logging.getLogger',
-)  # type: Union[str, Callable[[str], Logger]]
 
-# Allow per-queue overrides of the backend.
-BACKEND_OVERRIDES = setting('BACKEND_OVERRIDES', {})  # type: Mapping[QueueName, str]
+class AppSettings:
```

Comment:
Sorry it's taken a while to get back to this. Are you suggesting that `add_layer` should be part of the public interface? I had avoided doing that since the consequences of calling it arbitrarily aren't well defined. That said, this is meant to be internal to this package anyway, so maybe it's ok? I'll do that and add a comment clarifying that this file is internal API.

Reply:
80d2fd4ab53e1dd60702d8a76fd81bb8c41d8141